File: generator.dart
import 'dart:io';

class Generator {
  static const String CSV_LIST_SEPARATOR = '|';
  static const String ENUM_CLASS_NAME = "ErrorEnum";
  static const String FILE_SUFFIX_NAME = "dart";
  String inputFile;
  String outputDir;
  List<String> attributeList;

  Generator(String inputFile, String outputDir) {
    this.inputFile = inputFile;
    this.outputDir = outputDir;
  }

  void generate() {
    Map<String, Map<String, String>> dataMap = obtainDataMap();
    generateConstantFile(dataMap);
    generateEnumFile(dataMap);
  }

  void generateConstantFile(Map<String, Map<String, String>> dataMap) {
    for (String attribute in attributeList) {
      String className = toBigHump(attribute);
      String filePath = obtainOutputFilePath(className);
      String content = obtainConstantTemplate(className, attribute, dataMap);
      generateFile(filePath, content);
    }
  }

  void generateEnumFile(Map<String, Map<String, String>> dataMap) {
    String className = ENUM_CLASS_NAME;
    String filePath = obtainOutputFilePath(className);
    String content = obtainEnumTemplate(className, dataMap);
    generateFile(filePath, content);
  }

  Map<String, Map<String, String>> obtainDataMap() {
    var dataList = List<List<String>>();
    List<String> rowList = parseFile(inputFile);
    for (String row in rowList) {
      dataList.add(splitRowData(row));
    }
    attributeList = dataList.removeAt(0);

    var dataMap = Map<String, Map<String, String>>();
    var attributeSize = attributeList.length;
    for (var rowDataList in dataList) {
      fillList(rowDataList, attributeSize);
      String name = obtainLabelName(rowDataList[0]);
      var itemMap = Map<String, String>();
      dataMap[name] = itemMap;

      for (int index = 0; index < attributeSize; index++) {
        String attribute = attributeList[index];
        itemMap[attribute] = rowDataList[index];
      }
    }
    return dataMap;
  }

  List<String> splitRowData(String row) {
    List<String> list = List<String>();
    var rowArr = row.split(CSV_LIST_SEPARATOR);
    for (String data in rowArr) {
      list.add(data.trim());
    }
    return list;
  }

  List<String> fillList(List<String> list, int tagSize) {
    if (list.length < tagSize) {
      for (int index = 0; index < tagSize - list.length; index++) {
        list.add("");
      }
    }
    return list;
  }

  String obtainLabelName(String name) {
    String regex = "\\s+";
    return name.trim().replaceAll(RegExp(regex), "_").toUpperCase();
  }

  String obtainPropertyName(String property) {
    return toSmallHump(property);
  }

  String toBigHump(String str) {
    return toHump(str, false);
  }

  String toSmallHump(String str) {
    return toHump(str, true);
  }

  String toHump(String str, bool isSmallHump) {
    var builder = StringBuffer();
    var strArr = str.split("[ _-]");
    String subStr = "";
    for (int index = 0; index < strArr.length; index++) {
      subStr = strArr[index];
      if (index != 0 || !isSmallHump) {
        subStr = capitalize(subStr);
      }
      builder.write(subStr);
    }
    return builder.toString();
  }

  String capitalize(String str) {
    return str.substring(0, 1).toUpperCase() + str.substring(1);
  }

  String obtainConstantTemplate(String className, String attribute, Map<String, Map<String, String>> dataMap) {
    var builder = StringBuffer();
    builder.write("class $className {\n");
    for (MapEntry<String, Map<String, String>> data in dataMap.entries) {
      String format = "\tstatic const ${data.key} = \"${data.value[attribute]}\";\n";
      builder.write(format);
    }
    builder.write("}\r\n");
    return builder.toString();
  }

  String obtainEnumTemplate(String className, Map<String, Map<String, String>> dataMap) {
    return "class $className {\n" +
        "${obtainEnumDataList(dataMap, className)}\n" +
        "${obtainEnumProperty()}\n" +
        "${obtainEnumConstructor(className)}\n" +
        "}";
  }

  String obtainEnumDataList(Map<String, Map<String, String>> dataMap, String className) {
    var builder = StringBuffer();
    List<String> propertyList = obtainEnumPropertyList();
    var nameList = dataMap.keys.toList();
    for (int index = 0; index < nameList.length; index++) {
      String name = nameList[index];
      var itemMap = dataMap[name];
      String propertyArrString = propertyList.map((property) => "\"${itemMap[property]}\"").join(", ");
      builder.write("\tstatic const $name = ${obtainPrivateConstructorName(className)}($propertyArrString);\n");
    }
    return builder.toString();
  }

  String obtainEnumConstructor(String className) {
    var builder = StringBuffer();
    List<String> propertyList = obtainEnumPropertyList();
    String propertyArrString = propertyList.map((property) => "this.${obtainPropertyName(property)}").join(", ");
    builder.write("\tconst ${obtainPrivateConstructorName(className)}($propertyArrString);\n");
    return builder.toString();
  }

  String obtainPrivateConstructorName(String className) {
    return "$className._init";
  }

  String obtainEnumProperty() {
    var builder = StringBuffer();
    List<String> propertyList = obtainEnumPropertyList();
    for (String property in propertyList) {
      builder.write("\tfinal ${obtainPropertyName(property)};\n");
    }
    return builder.toString();
  }

  List<String> obtainEnumPropertyList() {
    var list = List<String>();
    for (int index = 1; index < attributeList.length; index++) {
      String tag = attributeList[index];
      list.add(tag);
    }
    return list;
  }

  String obtainOutputFilePath(String name) {
    return "$outputDir${Platform.pathSeparator}$name.$FILE_SUFFIX_NAME";
  }

  void generateFile(String filePath, String content) {
    try {
      File(filePath).create(recursive: true).then((file) {
        file.writeAsString(content);
      });
    } catch (e) {
      print(e);
    }
  }

  List<String> parseFile(String filePath) {
    var list = List<String>();
    try {
      File file = File(filePath);
      list = file.readAsLinesSync();
    } catch (e) {
      print(e);
    }
    return list;
  }
}

void main() {
  var currentPath = Directory.current.path;
  String inputFile = currentPath + Platform.pathSeparator + "error-code.csv";
  String outputDir = currentPath + Platform.pathSeparator + "output" + Platform.pathSeparator + "dart";
  Generator(inputFile, outputDir).generate();
}

File: generator.py
# coding=UTF-8

import sys
import os
import codecs
import re

FILE_CODING = 'utf-8' 
CSV_LIST_SEPARATOR = '|'
ENUM_FILE_NAME = "error_enum";
FILE_SUFFIX_NAME = "py";

class Generator:
    def __init__(self, input_file=None, output_dir=None):
        if output_dir is None:
            output_dir = os.path.split(os.path.realpath(__file__))[0]
        self.input_file = input_file
        self.output_dir = output_dir
        self.attribute_list = []
        self.templates = []

    def generate(self):
        data_map = self.obtain_data_map()
        self.generate_constant_file(data_map)
        self.generate_enum_file(data_map)

    def generate_constant_file(self, data_map):
        for attribute in self.attribute_list:
            name = self.to_underline(attribute)
            output_file = self.obtain_output_file_path(name)
            content = self.obtain_constant_template(name, attribute, data_map)
            self.generate_file(content, output_file)

    def generate_enum_file(self, data_map):
        name = ENUM_FILE_NAME
        output_file = self.obtain_output_file_path(name)
        content = self.obtain_enum_template(name, data_map)
        self.generate_file(content, output_file)

    def obtain_data_map(self):
        data_list = []
        row_list = self.parse_file()
        for row in row_list:
            data_list.append(self.split_row_data(row))

        self.attribute_list = data_list.pop(0)
        data_map = {}
        attribute_size = len(self.attribute_list)
        for row_data_list in data_list:
            row_data_list = self.fill_list(row_data_list, attribute_size)
            name = self.obtain_label_name(row_data_list[0])
            data_map[name] = {}
            for index in range(attribute_size):
                attribute = self.attribute_list[index]
                data_map[name][attribute] = row_data_list[index]
        return data_map

    def split_row_data(self, row): 
        list = []
        row_arr = row.strip().split(CSV_LIST_SEPARATOR)
        for data in row_arr:
            list.append(data.strip())
        return list

    def fill_list(self, list, size):
        if len(list) < size:
            for index in range(size - len(list)):
                list.append("")
        return list

    def obtain_label_name(self, name):
        return re.sub('\\s+', '_', name.strip().upper())

    def to_underline(self, str):
        return str

    def obtain_constant_template(self, name, attribute, data_map):
        template = 'class {name}:\n'.format(name=name)
        for key, value in data_map.items():
            template += '\t{key} = \'{value}\'\n'.format(key=key, value=value[attribute])
        template += '\r\n'
        return template

    def obtain_enum_template(self, name, data_map):
        return '{import_package}\n' \
            '{enum_property}\n' \
            '@unique\nclass {name}(Enum):\n' \
            '{data_list}\n' \
            '{property_getter}' \
                .format(import_package=self.obtain_enum_import_package(), enum_property=self.obtain_enum_property(), name=name, data_list=self.obtain_enum_data_list(data_map), property_getter=self.obtain_enum_property_getter())

    def obtain_enum_data_list(self, data_map):
        template = ''
        property_list = self.obtain_enum_property_list()
        name_list = list(data_map.keys())
        namedtuple = self.obtain_enum_namedtuple()
        for index in range(len(name_list)):
            name = name_list[index]
            property_arr = ", ".join('\'{}\''.format(data_map[name][property]) for property in property_list)
            template += '\t{name} = {namedtuple}({property_arr})\n' \
                .format(name=name, namedtuple=namedtuple, property_arr=property_arr)
        return template

    def obtain_enum_property(self):
        namedtuple = self.obtain_enum_namedtuple()
        property_list = self.obtain_enum_property_list()
        property_arr = ', '.join('\'{}\''.format(property) for property in property_list)
        return '{namedtuple} = namedtuple(\'data\', [{property_arr}])\n'.format(namedtuple=namedtuple, property_arr=property_arr)

    def obtain_enum_namedtuple(self):
        return 'Tuple'

    def obtain_enum_import_package(self):
        return 'from enum import Enum, unique\n' \
            'from collections import namedtuple\n'

    def obtain_enum_property_getter(self):
        template = ''
        property_list = self.obtain_enum_property_list()
        for property in property_list:
            template += '\t@property\n' \
            '\tdef {property}(self):\n' \
            '\t\treturn self.value.{property}\n\n'.format(property=property)
        return template

    def obtain_enum_property_list(self):
        return self.attribute_list
        
    def obtain_output_file_path(self, name):
        self.pre_generate_dir(self.output_dir)
        return '{directory}{separator}{name}.{suffix}' \
            .format(directory=self.output_dir, separator=os.sep, name=name, suffix=FILE_SUFFIX_NAME)

    def pre_generate_dir(self, directory):
        if not os.path.exists(directory):
            os.mkdir(directory)

    def generate_file(self, content, file):
        try:
            writer = open(file, 'w')
            writer.write(content)
            writer.close()
        except Exception as e:
            print(e)

    def parse_file(self):
        try:
            reader = codecs.open(self.input_file, 'r', FILE_CODING)
            data_list = []
            for line in reader:
                data_list.append(line)
            reader.close()
        except FileNotFoundError:
            print('file not found')
        except Exception as e:
            print(e)
        return data_list

if __name__ == '__main__':
    current_path = os.path.dirname(sys.argv[0])
    input_file = os.path.join(current_path,'error-code.csv')
    output_dir = os.path.join(current_path, 'output', 'python')
    Generator(input_file, output_dir).generate()
